numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
+            # accept objects exposing either old-style or new-style buffer interface
+            try:
+                o = memoryview(o)
+            except TypeError:
+                o = buffer(o)
+
+        else:  # pragma: py2 no cover
+            o = memoryview(o)
+
+        # N.B., this is not documented, but np.array() will accept an object exposing
+        # a buffer interface, and will take a view of the memory rather than making a
+        # copy, preserving type information where present
+        o = np.array(o, copy=False)
+
+    # check for object arrays
+    if o.dtype == object:
+        raise ValueError('object arrays are not supported')
+
+    if flatten:
+
+        # flatten the array to 1 dimension
+        o = o.reshape(-1, order='A')
+
+    return o
+
+
+def ensure_memoryview(o, flatten=True):
+    """Obtain a :class:`memoryview` with a view of memory exposed by `o`.
+
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
+
+    Returns
+    -------
+    o : memoryview
+
+    """
+
+    # go via numpy, for convenience
+    o = ensure_ndarray_from_memory(o, flatten=flatten)
+
+    # check for datetime or timedelta ndarray, cannot take a memoryview of those
+    if o.dtype.kind in 'Mm':
+        o = o.view(np.int64)
+
+    # expose as memoryview
+    o = memoryview(o)
+
+    return o
+
+
+def ensure_bytes(o):
+    """Obtain a bytes object from memory exposed by `o`."""
+
+    if not isinstance(o, binary_type):
+
+        # obtain memoryview
+        m = ensure_memoryview(o)
+
+        # create bytes from memory
+        o = m.tobytes()
+
+    return o
+
+
+if PY2:  # pragma: py3 no cover
+    # Under PY2 some codecs are happier if they are provided with a buffer object
+    # rather than a memoryview, so here we provide a convenience function to obtain a
+    # buffer object from a range of possible inputs.
+
+    def ensure_buffer(o):
+        """Obtain a :class:`buffer` with a view of memory exposed by `o`.
+
+        Parameters
+        ----------
+        o : bytes-like
+            Any object exposing a memory buffer. Can be an object exposing either
+            old-style or new-style buffer interface.
+
+        Returns
+        -------
+        o : buffer
+
+        """
+
+        # go via numpy, for convenience
+        o = ensure_ndarray_from_memory(o)
 
+        # expose as buffer
+        o = buffer(o)
```

Comment:
Already mentioned above, but would suggest using `np.getbuffer` as well to ensure write permissions are preserved.

Though much like `memoryview` I think it is easier to just always give an `ndarray` instead since it supports both protocols. That avoids getting into too much code that is different between Python 2/3 and is generally much more functional for end users.

This is just my opinion though. Feel free to disagree. Would be interested in hearing your thoughts on this as well.

Reply:
FWIW the only codec that actually needs the ``ensure_buffer()`` function is ``ZLib``. 

Also FWIW the reason why I didn't just use ``ensure_contiguous_ndarray()`` everywhere, and get rid of ``ensure_memoryview()`` and ``ensure_buffer()``, is because of arrays with datetime or timedelta data types. Codecs like ``ZLib`` and ``BZ2`` and anything that will try to access data via the new-style buffer interface in Python 3 will error on an array with datetime or timedelta datatype.

I could push the conversion from datetime/timedelta to integer datatype up inside ``ensure_contiguous_ndarray()``, which would mean that that single function could be used everywhere, and both ``ensure_memoryview()`` and ``ensure_buffer()`` could be deleted. Seems a bit beyond the scope of just ensuring contiguous memory, but it would simplify the code.

Reply:
Yeah, I dealt with the same tension regarding the scope of the `ndarray` function. Though increasingly I think we should just move the datetime/timedelta code there and think of a better function name.

Reply:
Yeah, maybe ``gimme_an_ndarray_with_a_C_contiguous_memory_buffer_that_supports_the_new_style_buffer_protocol_and_doesnt_have_an_object_dtype()`` :smiley:. Seriously though, that is the essence of the function, so some name that captures that would be good, and I'd be happy to move the datetime/timedelta code in.

Reply:
Yeah, it starts to feel like we are writing Java code. 😄

Having `serializable` in there sounds really good for some reason. 😉